csharp/src/Ice/Properties.cs

```diff
@@ -164,17 +400,443 @@ public interface Properties
     ///  <param name="options">The command-line options.
     ///  </param>
     /// <returns>The command-line options that do not start with one of the listed prefixes, in their original order.</returns>
-    string[] parseIceCommandLineOptions(string[] options);
+    public string[] parseIceCommandLineOptions(string[] options)
+    {
+        string[] args = options;
+        foreach (var name in Ice.Internal.PropertyNames.clPropNames)
+        {
+            args = parseCommandLineOptions(name, args);
+        }
+        return args;
+    }
 
     /// <summary>
     /// Load properties from a file.
     /// </summary>
     /// <param name="file">The property file.</param>
-    void load(string file);
+    public void load(string file)
+    {
+        try
+        {
+            using StreamReader sr = new StreamReader(file);
+            parse(sr);
+        }
+        catch (IOException ex)
+        {
+            var fe = new FileException(ex);
+            fe.path = file;
+            throw fe;
+        }
+    }
 
     /// <summary>
     /// Create a copy of this property set.
     /// </summary>
     /// <returns>A copy of this property set.</returns>
-    Properties ice_clone_();
+    public Properties ice_clone_()
+    {
+        lock (this)
+        {
+            return new Properties(this);
+        }
+    }
+
+    public List<string> getUnusedProperties()
+    {
+        lock (this)
+        {
+            var unused = new List<string>();
+            foreach (KeyValuePair<string, PropertyValue> entry in _properties)
+            {
+                if (!entry.Value.used)
+                {
+                    unused.Add(entry.Key);
+                }
+            }
+            return unused;
+        }
+    }
+
+    internal Properties(Properties p) : this()
```

Comment:
This is not proper C# code.

Reply:
This constructor is only used in `ice_clone_`

```
    public Properties ice_clone_()
    {
        lock (this)
        {
            return new Properties(this);
        }
    }
```

We can do the cloning in `ice_clone_` and remove this constructor. Can we rename `ice_clone_` to `Clone`?

Reply:
> Can we rename ice_clone_ to Clone?

I think the Ice 3.8 C# API should remain as close as possible to the 3.7 C# API and this would be a gratuitous renaming.

If you're willing to refactor your code with lots of new names, then port your code to IceRPC C#.